File: 059-对称的二叉树/059-isSymmetrical.py
# -*- coding:utf-8 -*-
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None
class Solution:
    def isSymmetrical(self, pRoot):
        # write code here
        if pRoot is None:
            return True
        return self.subTreeIsSymmetrical(pRoot.left,pRoot.right)
    def subTreeIsSymmetrical(self,lroot,rroot):
        if lroot is None and rroot is None:
            return True
        if lroot and rroot and lroot.val==rroot.val:
            return self.subTreeIsSymmetrical(lroot.left,rroot.right) and self.subTreeIsSymmetrical(lroot.right,rroot.left)
        return False

File: 059-对称的二叉树/059-isSymmetrical.cpp
/*
struct TreeNode {
    int val;
    struct TreeNode *left;
    struct TreeNode *right;
    TreeNode(int x) :
            val(x), left(NULL), right(NULL) {
    }
};
*/
class Solution {
public:
    bool isSymmetrical(TreeNode* pRoot)
    {
        if(pRoot==nullptr){
            return true;
        }
        return subTree(pRoot->left,pRoot->right);
    }
    bool subTree(TreeNode* rleft,TreeNode* rright){
        if(rleft==nullptr && rright==nullptr){
            return true;
        }
        if(rleft != nullptr && rright!=nullptr && rleft->val==rright->val){
            return subTree(rleft->right,rright->left) && subTree(rleft->left,rright->right);
        }
        return false;
    }

};
